File: Dynamic Programming/1277. Count Square Submatrices with All Ones.py
class Solution:
    def countSquares(self, matrix: List[List[int]]) -> int:
        if not matrix:
            return 0
        m,n = len(matrix), len(matrix[0])
        dp = [[0 for _ in range(n)] for _ in range(m)]
        res = 0
        for i, row in enumerate(matrix):
            for j, k in enumerate(row):
                if k==0:
                    continue 
                dp[i][j] = 1
                if i>0 and j > 0 and matrix[i-1][j] == matrix[i][j-1] == matrix[i-1][j-1] == 1:
                        dp[i][j] +=  min(dp[i-1][j-1],dp[i][j-1],dp[i-1][j])
                res += dp[i][j]
        return res

#TOP-Down DP


class Solution:
    def countSquares(self, matrix: List[List[int]]) -> int:
        m, n = len(matrix), len(matrix[0])

        @functools.lru_cache(None) # use this for memory search
        def dp(i,j):
            if matrix[i][j]==1 : 
                if i!=0 and j!=0:
                    return min(dp(i-1,j), dp(i,j-1), dp(i-1,j-1))+1
                else:
                    return 0+1
            else:
                return 0
        
        return sum(dp(i,j) for i in range(m) for j in range(n))

File: Dynamic Programming/1277. Count Square Submatrices with All Ones.cpp
class Solution {
public:
    int countSquares(vector<vector<int>>& matrix) {
        int res = 0;
        for (int i = 0; i < matrix.size(); ++i)
            for (int j = 0; j < matrix[0].size(); res += matrix[i][j++])
                if (matrix[i][j] && i && j)
                    matrix[i][j] += min({matrix[i - 1][j - 1], matrix[i - 1][j], matrix[i][j - 1]});
        return res;
    }
};

/**
 matrix[i][j] += matrix[i - 1][j - 1]; 不可以这么写的原因比如:
 下面的 matrix[2][2] = matrix[1][1] + 1 = 3, 但实际长度是2

             DP will be 
   1, 1, 0    1, 1, 0
   1, 1, 1    1, 2, 1
   0, 1, 1    0, 1, 0      

 */


class Solution {
public:
    int countSquares(vector<vector<int>>& matrix) {
        for (int i = 1; i < matrix.size(); ++i) {
            for (int j = 1; j < matrix[0].size(); ++j) {
                if (!matrix[i][j]) {
                    continue;
                }
                int l = min(matrix[i - 1][j], matrix[i][j - 1]);
                matrix[i][j] = matrix[i - l][j - l] ? l + 1 : l;
            }
        }
        int result = 0;
        for (const auto& row : matrix) {
            for (const auto& x : row) {
                result += x;
            }
        }
        return result;
    }
};